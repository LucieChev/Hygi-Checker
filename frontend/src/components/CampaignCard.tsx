export type CampaignType = {
  title: string
  url: string
  urlNumber: number
}

type CampaignCardProps = {
  campaign: CampaignType
}

export default function CampaignCard({
  campaign: { title, url, urlNumber },
}: CampaignCardProps) {
  return (
    <div className="[w-400] flex flex-column p-2 gap-2 rounded-lg border">
      <h2>{title}</h2>
      <div className="flex flex-column gap-1">
        <p>{url}</p>
        <p>{urlNumber}</p>
      </div>
    </div>
  )
}
