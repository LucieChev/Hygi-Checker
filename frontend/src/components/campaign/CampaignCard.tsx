"use client";

import Image from "next/image";
import { Badge } from "../ui/badge";
import { Button } from "../ui/button";
import {
  Card,
  CardTitle,
  CardDescription,
  CardContent,
  CardFooter,
} from "../ui/card";
import { Switch } from "../ui/switch";
import Link from "next/link";
import { CampaignCardProps } from "@/types/interfaces";
import { useState, useEffect } from "react";
import { useToast } from "@/components/ui/use-toast";
import {
  useCountUrlFromCampaignQuery,
  useSwitchWorkingCampaignMutation,
  useCampaignsByUserIdQuery,
} from "@/types/graphql";
import { testPerDay } from "@/utils/chartFunction/testsPerDay";
import { formatDate } from "@/utils/chartFunction/formatDate";

export default function CampaignCard({ data }: CampaignCardProps) {
  const { toast } = useToast();
  const [isWorkingCampaign, setIsWorkingCampaign] = useState<boolean>(
    data?.isWorking ?? false
  );

  const { refetch } = useCampaignsByUserIdQuery();

  const [switchWorkingCampaignMutation] = useSwitchWorkingCampaignMutation({
    onCompleted: (data) => {
      toast({
        title: data.switchWorkingCampaign.message,
        variant: `${isWorkingCampaign === false ? "default" : "success"}`,
      });
      refetch();
    },
    onError: () => {
      toast({
        title: `Something went wrong. Please try again`,
        variant: "destructive",
      });
    },
  });

  const handleIsWorking = () => {
    setIsWorkingCampaign(!isWorkingCampaign);
  };

  const { data: countUrl } = useCountUrlFromCampaignQuery({
    variables: {
      campaignId: data?.id,
    },
  });

  const count = countUrl?.countUrlFromCampaign.count;
  const testEachDay = testPerDay(count as number, data?.intervalTest as number);

  useEffect(() => {
    if (isWorkingCampaign !== data?.isWorking) {
      switchWorkingCampaignMutation({
        variables: {
          input: {
            campaignId: data?.id,
            isWorking: isWorkingCampaign,
          },
        },
      });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isWorkingCampaign]);

  return (
    <Card className="flex flex-col w-[350px] h-[350px]">
      <Image
        src={data?.image ?? "../../../public/logo_large.svg"}
        alt="image"
        className="w-full rounded-t-lg"
        width={350}
        height={200}
        priority
      />

      <CardContent className="flex justify-between p-6 items-center">
        <div className="flex flex-col">
          <CardTitle className="text-2xl">{data.name}</CardTitle>
          <CardDescription>
            {formatDate(new Date(data?.createdAt))}
          </CardDescription>
        </div>
        <Switch checked={isWorkingCampaign} onCheckedChange={handleIsWorking} />
      </CardContent>
      <CardFooter className="flex justify-between">
        <div>
          <Badge>{count ?? 0} urls</Badge>
          <Badge className="ml-2">{testEachDay ?? 0} tests / day</Badge>
        </div>

        <Link href={`/dashboard/campaign/details/${data.id}`}>
          <Button data-testid="watch-button" variant="outline">
            Details
          </Button>
        </Link>
      </CardFooter>
    </Card>
  );
}
