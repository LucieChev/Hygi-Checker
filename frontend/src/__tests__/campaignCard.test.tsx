import CampaignCard from "@/components/dashboard/CampaignCard";
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";

describe("CampaignCard component", () => {
  it("renders correctly", () => {
    const view = render(
      <CampaignCard
        campaign={{
          title: "teeeest",
          url: "urllll",
          urlNumber: 4,
        }}
      />
    );
    expect(screen.getByText(/teeeest/)).toBeInTheDocument();
    expect(screen.getByText(/4/)).toBeInTheDocument();
    expect(view.baseElement).toMatchSnapshot();
  });
});
