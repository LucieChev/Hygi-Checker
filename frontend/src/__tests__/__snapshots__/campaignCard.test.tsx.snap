// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CampaignCard component renders correctly 1`] = `
<body>
  <div>
    <div
      class="rounded-lg bg-white text-card-foreground shadow-sm flex flex-col w-[350px] h-[350px]"
    >
      <img
        alt="image"
        class="w-full rounded-t-lg"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="200"
        src="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1564406836777-5964b5c6c3a5%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DMnwxfDB8MXxyYW5kb218MHx8d2FsbHBhcGVyc3x8fHx8fDE3MTU4NzgwNDQ%26ixlib%3Drb-4.0.3%26q%3D80%26utm_campaign%3Dapi-credit%26utm_medium%3Dreferral%26utm_source%3Dunsplash_source%26w%3D1080&w=750&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1564406836777-5964b5c6c3a5%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DMnwxfDB8MXxyYW5kb218MHx8d2FsbHBhcGVyc3x8fHx8fDE3MTU4NzgwNDQ%26ixlib%3Drb-4.0.3%26q%3D80%26utm_campaign%3Dapi-credit%26utm_medium%3Dreferral%26utm_source%3Dunsplash_source%26w%3D1080&w=384&q=75 1x, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1564406836777-5964b5c6c3a5%3Fcrop%3Dentropy%26cs%3Dtinysrgb%26fit%3Dmax%26fm%3Djpg%26ixid%3DMnwxfDB8MXxyYW5kb218MHx8d2FsbHBhcGVyc3x8fHx8fDE3MTU4NzgwNDQ%26ixlib%3Drb-4.0.3%26q%3D80%26utm_campaign%3Dapi-credit%26utm_medium%3Dreferral%26utm_source%3Dunsplash_source%26w%3D1080&w=750&q=75 2x"
        style="color: transparent;"
        width="350"
      />
      <div
        class="flex justify-between p-6 items-center"
      >
        <div
          class="flex flex-col"
        >
          <h3
            class="font-semibold tracking-tight text-2xl"
          >
            Campaign1_User1
          </h3>
          <p
            class="text-sm text-muted-foreground"
          >
            2024-07-14
          </p>
        </div>
        <button
          aria-checked="true"
          class="peer inline-flex h-6 w-11 shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-secondary data-[state=unchecked]:bg-input"
          data-state="checked"
          role="switch"
          type="button"
          value="on"
        >
          <span
            class="pointer-events-none block h-5 w-5 rounded-full bg-background shadow-lg ring-0 transition-transform data-[state=checked]:translate-x-5 data-[state=unchecked]:translate-x-0"
            data-state="checked"
          />
        </button>
      </div>
      <div
        class="items-center p-6 pt-0 flex justify-between"
      >
        <div>
          <div
            class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-primary text-primary-foreground hover:bg-primary/80"
          >
            0
             urls
          </div>
          <div
            class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-primary text-primary-foreground hover:bg-primary/80 ml-2"
          >
            NaN
             tests / day
          </div>
        </div>
        <a
          href="/dashboard/campaign/details/1"
        >
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-secondary hover:bg-secondary/90 hover:text-accent-foreground h-10 px-4 py-2"
            data-testid="watch-button"
          >
            Details
          </button>
        </a>
      </div>
    </div>
  </div>
</body>
`;
